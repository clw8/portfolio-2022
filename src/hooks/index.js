export { default as useScrollTo } from "./use-scroll-to.js"
export { default as useModal } from "./use-modal.js"
