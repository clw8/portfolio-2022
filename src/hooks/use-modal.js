import * as React from "react"
import Reveal from 'react-reveal';

const useModal = (children) => {
    const [modalShown, setModalShown] = React.useState(false)
    const [modalChildren, setModalChildren] = React.useState(children) // children must be a function
    const [initialAnimationPlayState, setInitialAnimationPlayState] = React.useState(false) // children must be a function
    
    if (children && React.isValidElement) {
        console.warn("children must be declared in the modal as a function")
    }

    const preventScroll = () => {
        document.body.style.overflow = "hidden"
      }

    const enableScroll = () => {
        document.body.style.overflow = ""
      }

    const openModal = (children) => {
        modalShown ? enableScroll() : preventScroll() 
        setModalShown(!modalShown)
        !initialAnimationPlayState && setInitialAnimationPlayState(true)
        
        if (children) {
            if (!React.isValidElement) console.warn("children must be declared in the modal as a function")
            else setModalChildren(children) // children must be a function
        }
    }

    const modalClass = `modal ${modalShown ? "modal__enter" : (initialAnimationPlayState ? "modal__leave" : "")}`

    const renderModal = (
        <div className={modalClass}>
            <div className="modal__inner">
                <div className="modal__content">
                    {modalChildren}
                </div>
            </div>
        </div>
    )

    return [
        renderModal,
        modalShown,
        openModal,
    ]
}

export default useModal