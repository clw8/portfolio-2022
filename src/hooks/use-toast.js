import { useContext } from "react";
import { ToastContext } from "../context-providers/toast-provider"

const useToast = () => {
    const {
        toastIsShown, setShowToast, toastType, setToastType, toastTypes, toastMessage, setToastMessage
    } = useContext(ToastContext)

    const showErrorToast = (message = "") => {
        setShowToast(true)
        setToastMessage(message)
        setToastType(toastTypes.ERROR)
    }

    const showSuccessToast = (message = "") => {
        setShowToast(true)
        setToastMessage(message)
        setToastType(toastTypes.SUCCESS)
    }

    const hideToast = () => {
        setShowToast(false)
        setToastMessage("")
    }

    return {
    toastIsShown, toastType, toastTypes, toastMessage, hideToast, showErrorToast,  showSuccessToast
    }
}

export default useToast
