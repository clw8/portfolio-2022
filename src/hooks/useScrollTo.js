import React, { useState } from "react";

const useScrollTo = () => {

    const onClickHashScrollTo = (event) => {
        event.preventDefault();
        let hash = event.target.hash

        if (hash === "/") {
          document.body.scrollIntoView(true);
        } else {
          document.querySelector(hash).scrollIntoView({block:"start", behavior:"smooth"})
        }
        
      }

      return {
          onClickHashScrollTo,
      }
}

export default useScrollTo