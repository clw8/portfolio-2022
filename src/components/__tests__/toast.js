import {cleanup, fireEvent, render, act } from '@testing-library/react';
import "@testing-library/jest-dom"
import React, { useEffect } from "react"

import ToastProvider from '../../context-providers/toast-provider';
import { useToast } from '../../hooks';

const Layout = ({children}) => {
    return (
        <div style={{minHeight: "100vh", position: "relative"}}>
            <ToastProvider>
                {children}
            </ToastProvider>
        </div>
    )
}
let showToast;
const Page = () => {
    const { showSuccessToast } = useToast()
    showToast = () => showSuccessToast("hi there!")

    return <div style={{ height: "1400px" }}>Home page</div>
}

describe("Toast", () => {
    let toast, asFragmentVariable;
    beforeEach(() => {
        const {getByRole, asFragment} = render(
            (<Layout>
                <Page />
            </Layout>))

        act(() => {
            showToast()
            toast = getByRole("alert")
            asFragmentVariable = asFragment
        })
    })

    it ("renders correctly", () => {
        expect(toast).toBeInTheDocument()
        const firstRender = asFragmentVariable()
        expect(firstRender).toMatchSnapshot()
    })

    it("is triggered successfully", () => {
        act(() => {
            fireEvent.animationEnd(toast)
        })

        expect(toast).toHaveAttribute("animation", "toast-in-end")

        expect(toast).toBeVisible()
        expect(toast).toHaveTextContent("hi there!")
    })
    
    it("hides toast on user click", () => {
        fireEvent.animationEnd(toast)
        act(() => {
            fireEvent.click(toast)
            fireEvent.animationEnd(toast)
        })

        expect(toast).not.toHaveAttribute("animation", "toast-out-end")
        expect(toast).not.toHaveTextContent("hi there!")
    })

})