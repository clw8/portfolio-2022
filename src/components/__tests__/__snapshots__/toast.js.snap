// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders correctly 1`] = `
<DocumentFragment>
  <div
    style="min-height: 100vh; position: relative;"
  >
    <div
      style="height: 1400px;"
    >
      Home page
    </div>
    <div
      animation="toast-in"
      class="toast toast--success"
      role="alert"
    >
      <p>
        hi there!
      </p>
    </div>
  </div>
</DocumentFragment>
`;
